### PR TITLE
feat(alerting): Add group-specific WebHook URL for Discord

### PR DESCRIPTION
**Pull request by appleboy**
Add group-specific webhook URL for discord alert

Provides support for paging multiple Discords based on the group selector while keeping backward compatibility to the old Discords configuration manifest

integration per team can be specified in the overrides sections in an array form.

```yaml
alerting:
  discord: 
    webhook-url: "https://discord.com/api/webhooks/**********/**********"
    # You can also add group-specific to keys, which will 
    # override the to key above for the specified groups
    overrides:
      - group: "core"
        webhook-url: "https://discord.com/api/webhooks/************"

endpoints:
  - name: website
    url: "https://twin.sh/health"
    interval: 30s
    conditions:
      - "[STATUS] == 200"
      - "[BODY].status == UP"
      - "[RESPONSE_TIME] < 300"
    alerts:
      - type: discord
        enabled: true
        description: "healthcheck failed"
        send-on-resolved: true

  - name: back-end
    group: core
    url: "https://example.org/"
    interval: 5m
    conditions:
      - "[STATUS] == 200"
      - "[CERTIFICATE_EXPIRATION] > 48h"
    alerts:
      - type: discord
        enabled: true
        description: "healthcheck failed"
        send-on-resolved: true
```

ref: #96

Signed-off-by: Bo-Yi Wu <appleboy.tw@gmail.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -372,10 +381,23 @@ endpoints:
...
+    url: "https://example.org/"
+    interval: 5m
+    conditions:
+      - "[STATUS] == 200"
+      - "[CERTIFICATE_EXPIRATION] > 48h"
+    alerts:
+      - type: discord
+        enabled: true
+        description: "healthcheck failed"
+        send-on-resolved: true
```
Comment (TwiN) [inline README.md:396]: I don't think there's a need to explicitly showcase the effect that `alerting.*.overrides` has on the configuration, as it will create unnecessary noise in the documentation.

File: README.md [hunk 2]
```diff
@@ -346,18 +347,26 @@ ignored.
...
 
 ```yaml
 alerting:
   discord: 
     webhook-url: "https://discord.com/api/webhooks/**********/**********"
+    # You can also add group-specific to keys, which will 
+    # override the to key above for the specified groups
+    overrides:
+      - group: "core"
+        webhook-url: "https://discord.com/api/webhooks/************"
```
Comment (TwiN) [inline README.md:369]: I don't think there's a need to explicitly show the `alerting.*.overrides` configuration, as it will create unnecessary noise in the documentation (as you can see, I did the same for `alerting.discord.default-alert`

File: README.md [hunk 3]
```diff
@@ -346,18 +347,26 @@ ignored.
...
+|:------------------------------------------ |:------------------------------------------------------------------------------------------ |:------------- |
+| `alerting.discord`                         | Configuration for alerts of type `discord`                                                 | `{}`          |
+| `alerting.discord.webhook-url`             | Discord Webhook URL                                                                        | Required `""` |
+| `alerting.discord.default-alert`           | Default alert configuration. <br />See [Setting a default alert](#setting-a-default-alert) | N/A           |
+| `alerting.discord.overrides`               | List of overrides that may be prioritized over the default configuration                   | `[]`          |
+| `alerting.discord.overrides[].group`       | Endpoint group for which the configuration will be overridden by this configuration        | `""`          |
+| `alerting.discord.overrides[].webhook-url` | Teams Webhook URL                                                                          | `""`          |
```
Comment (TwiN) [inline README.md:359]: Small mistake in the description here

**Review by TwiN — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (appleboy) [reply README.md]: https://github.com/TwiN/gatus/pull/271/commits/75c34d152d3adfa5cdb5e393ee45e8bdb3693b78

File: README.md (hunk 2 above)
Comment (appleboy) [reply README.md]: https://github.com/TwiN/gatus/pull/271/commits/75c34d152d3adfa5cdb5e393ee45e8bdb3693b78

File: README.md (hunk 1 above)
Comment (appleboy) [reply README.md]: https://github.com/TwiN/gatus/pull/271/commits/75c34d152d3adfa5cdb5e393ee45e8bdb3693b78

File: README.md [hunk 4]
```diff
@@ -76,8 +77,8 @@ Have any feedback or want to share your good/bad experience with Gatus? Feel fre
   - [Endpoint groups](#endpoint-groups)
   - [Exposing Gatus on a custom port](#exposing-gatus-on-a-custom-port)
   - [Badges](#badges)
-    - [Uptime](#uptime)
-    - [Response time](#response-time)
```
[the remark this answers is not in the document]
Comment (appleboy) [reply README.md]: Sorry, my bad. https://github.com/TwiN/gatus/pull/271/commits/6fd0cb39676f72a871af8df92ba540fbe6a2771c

**Comment by appleboy** [comment]
@TwiN All Done, Please help to review again.

File: README.md [hunk 5]
```diff
@@ -346,13 +346,16 @@ ignored.
...
+|:------------------------------------------ |:------------------------------------------------------------------------------------------ |:------------- |
+| `alerting.discord`                         | Configuration for alerts of type `discord`                                                 | `{}`          |
+| `alerting.discord.webhook-url`             | Discord Webhook URL                                                                        | Required `""` |
+| `alerting.discord.default-alert`           | Default alert configuration. <br />See [Setting a default alert](#setting-a-default-alert) | N/A           |
+| `alerting.discord.overrides`               | List of overrides that may be prioritized over the default configuration                   | `[]`          |
+| `alerting.discord.overrides[].group`       | Endpoint group for which the configuration will be overridden by this configuration        | `""`          |
+| `alerting.discord.overrides[].webhook-url` | Discord Webhook URL                                                                          | `""`          |
```
Comment (TwiN) [inline README.md:358]:
```suggestion
| `alerting.discord.default-alert`           | Default alert configuration. <br />See [Setting a default alert](#setting-a-default-alert) | N/A           |
| `alerting.discord.overrides`               | List of overrides that may be prioritized over the default configuration                   | `[]`          |
| `alerting.discord.overrides[].group`       | Endpoint group for which the configuration will be overridden by this configuration        | `""`          |
| `alerting.discord.overrides[].webhook-url` | Discord Webhook URL                                                                        | `""`          |
```

File: alerting/provider/discord/discord_test.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/TwiN/gatus/v3/test"
 )
 
-func TestAlertProvider_IsValid(t *testing.T) {
+func TestAlertDefaultProvider_IsValid(t *testing.T) {
```
Comment (TwiN) [inline alerting/provider/discord/discord_test.go:14]: Why is this renamed? We're still testing the method `IsValid` of the struct `AlertProvider` 🤔

File: alerting/provider/discord/discord.go [hunk 7]
```diff
@@ -17,17 +17,36 @@ type AlertProvider struct {
...
+	registeredGroups := make(map[string]bool)
+	if provider.Overrides != nil {
+		for _, override := range provider.Overrides {
+			if isAlreadyRegistered := registeredGroups[override.Group]; isAlreadyRegistered || override.Group == "" || len(override.WebhookURL) == 0 {
+				return false
+			}
+			registeredGroups[override.Group] = true
+		}
+	}
+
```
Comment (TwiN) [inline alerting/provider/discord/discord.go:42]:
```suggestion
```

**Review by TwiN — CHANGES_REQUESTED** [review]

File: alerting/provider/discord/discord_test.go (hunk 6 above)
Comment (appleboy) [reply alerting/provider/discord/discord_test.go]: https://github.com/TwiN/gatus/pull/271/commits/9b764cfce29c1231da423fb1d5b531c241cf4bba done

**Comment by appleboy** [comment]
@TwiN All Done.

File: alerting/provider/discord/discord_test.go [hunk 8]
```diff
@@ -156,3 +193,66 @@ func TestAlertProvider_GetDefaultAlert(t *testing.T) {
...
+				},
+			},
+			InputGroup:     "group",
+			ExpectedOutput: "http://example01.com",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.Name, func(t *testing.T) {
+			if got := tt.Provider.getWebhookURLForGroup(tt.InputGroup); got != tt.ExpectedOutput {
+				t.Errorf("AlertProvider.getToForGroup() = %v, want %v", got, tt.ExpectedOutput)
```
Comment (TwiN) [inline alerting/provider/discord/discord_test.go:254]:
```suggestion
				t.Errorf("AlertProvider.getWebhookURLForGroup() = %v, want %v", got, tt.ExpectedOutput)
```
One last small change

**Review by TwiN — CHANGES_REQUESTED** [review]

**Comment by appleboy** [comment]
@TwiN Done. Please review again.

**Review by TwiN — APPROVED** [review]
My bad, I missed the notification.
Thank you for the contribution!
